crates/ty_python_semantic/src/types/class.rs

```diff
@@ -3462,7 +3480,52 @@ impl KnownClass {
                     _ => {}
                 }
             }
+            KnownClass::Deprecated => {
+                let assigned_to = index
+                    .try_expression(ast::ExprRef::from(call_expression))
+                    .and_then(|expr| expr.assigned_to(db));
+
+                let containing_assignment = assigned_to.as_ref().and_then(|assigned_to| {
+                    match assigned_to.node(module).targets.as_slice() {
+                        [ast::Expr::Name(target)] => Some(index.expect_single_definition(target)),
+                        _ => None,
+                    }
+                });
+
+                // Parsing something of the form:
+                //
+                // @deprecated("message")
+                // @deprecated("message", caregory = DeprecationWarning, stacklevel = 1)
+                //
+                // "Static type checker behavior is not affected by the category and stacklevel arguments"
+                // so we only need the message and can ignore everything else. The message is mandatory,
+                // must be a string literal, and always comes first.
+                //
+                // See:
+                // * <https://typing.python.org/en/latest/spec/directives.html#deprecated>
+                // * <https://docs.python.org/3/library/warnings.html>
+                let [Some(message), ..] = overload.parameter_types() else {
+                    if let Some(builder) = context.report_lint(&MISSING_ARGUMENT, call_expression) {
+                        builder
+                            .into_diagnostic("Missing `message` argument of `warnings.deprecated`");
+                    }
+                    return;
+                };
 
+                if let Some(message_ty) = message.into_string_literal() {
+                    overload.set_return_type(Type::KnownInstance(KnownInstanceType::Deprecated(
+                        DeprecatedInstance::new(db, message_ty, containing_assignment),
+                    )));
+                } else {
+                    if let Some(builder) =
+                        context.report_lint(&INVALID_ARGUMENT_TYPE, call_expression)
+                    {
+                        builder.into_diagnostic(
+                            "`warnings.deprecated`'s message must be a string literal",
+                        );
+                    }
```

Comment:
It feels like this is maybe too strict? The stub allows message to be a `LiteralString`, but here, we only allow arbitrary `Literal["…"]` string-literals, but not the common supertype `LiteralString`. Since `warnings.deprecated` also has a runtime effect, it seems fine to allow calls, even if we can't extract the message?

See this ecosystem hit in `altair` as an example where we currently emit this new diagnostic: https://github.com/vega/altair/blob/d30d030a89e2e2e72c6f5a4d497968b7dfb743bf/altair/utils/deprecation.py#L82

I would probably allow `message` to be `LiteralString` here, and return an instance of `deprecated` with some generic message?

Reply:
I also think that this will be incorrect for very long strings because ty falls back to `LiteralString` for overlong strings:

https://github.com/astral-sh/ruff/blob/4aee0398cbed38b84e596b47a5e97e56aef5d725/crates/ty_python_semantic/src/types/infer.rs#L4964-L4968

Reply:
But I can make it more lenient if it's messing up actual ecosystem crates.

Reply:
Aha ok I have read more on LiteralString and I now understand and agree with this assessment.

Reply:
Fair, it's 4096 bytes (I assumed it would be much shorter)